### PR TITLE
Deprecate declaring multiple targets of the same type

### PR DESCRIPTION
**Pull request by antohaby**
Hi!
I want to add deprecation information about declaring multiple targets in Kotlin Multiplatform Projects.

YT issue about deprecation: https://youtrack.jetbrains.com/issue/KT-59316

I also would like to ask your advice on how to structure this information best and consistently.

### CODE REVIEW

**Comment by antohaby** [comment]
This is currently in a draft state. We might adjust it with more real use cases. 
I will search for some community projects where multiple targets are declared and I'll try to understand why they do that and how to fix it in the recommended way. If I find some interesting and repeating cases I will consider adding them to this doc.

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 1]
```diff
@@ -490,3 +490,67 @@ projects.
 **When do the changes take effect?**
 
 In Kotlin 1.9.0, a deprecation warning is introduced when the `android` name is used in Kotlin Multiplatform projects.
+
+<anchor name="declaring-multiple-targets"></anchor>
+**What's changed?**
```
Comment (dsavvinov) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:495]: Heading is missing

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 2]
```diff
@@ -490,3 +490,67 @@ projects.
...
+
+Declaring multiple targets of the same kind in one project is not recommended and will be forbidden in future releases.
+There are multiple ways of achieving same results with other Kotlin Gradle Plugin mechanisms. As part of Kotlin
```
Comment (dsavvinov) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:498]: I'd drop those two sentences, we're expanding on this in details below anyways

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 3]
```diff
@@ -490,3 +490,67 @@ projects.
...
+                api(project(":lib-jvm"))
+            }
+        }
+    }
+}
+```
+
+**When do the changes take effect?**
+
+In Kotlin 1.9.20, a deprecation warning is introduced when multiple targets of the same type is used in Kotlin Multiplatform projects.
```
Comment (dsavvinov) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:556]: Normally, versions where next steps of deprecation cycle take palce are expected to be described here. E.g.:

- In Kotlin 1.9.20, deprecation warning is reported
- Kotlin 2.0, error will be reported in such cases, causing your build to fail

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 4]
```diff
@@ -490,3 +490,67 @@ projects.
...
+        val functionalTest by compilations.creating
+
+        // functionalTest should see public and internal declarations from 'main' compilation now.
+        // it is possible to associate with 'test' compilation as well
+        functionalTest.associateWith(compilations.getByName("main"))
+    }
+}
+```
+
+Different projects can be configured like this:
```
Comment (dsavvinov) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:531]: Currently it reads as continutiation of the previous idea/case, but it's in fact the second case (and second approach to addressing this deprecation). Let's separate them properly and add a bit of a structure along the way -- I'd suggest using numbered sections, like:

```
1. Creating custom compilations

...

2. Extracting new target into a separate Gradle project

...
```

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 5]
```diff
@@ -490,3 +490,67 @@ projects.
...
+kotlin {
+    jvm {
+        // Creates new compilation and 'jvmFunctionalTest' source set associated with it
+        val functionalTest by compilations.creating
+
+        // functionalTest should see public and internal declarations from 'main' compilation now.
+        // it is possible to associate with 'test' compilation as well
+        functionalTest.associateWith(compilations.getByName("main"))
+    }
+}
```
Comment (dsavvinov) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:528]: I'd suggest using tables with "before"/"after" because currently it's not clear for which multi-target configuration this code snippet is suggested.

Also, with separate target people might've had:
- separate compiler options
- separate dependencies
- separate sources

This is not the exhaustive list, but I think it covers most of the top-level requests (feel free to add something). Please show how to migrate them to the "separate compilation"-approach.

**Review by dsavvinov — CHANGES_REQUESTED** [review]

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md (hunk 2 above)
Comment (antohaby) [reply docs/topics/multiplatform/multiplatform-compatibility-guide.md:498]: But I think something should be placed under "what's changed" section.
I'll try to rephrase it to something like:

> As part of Kotlin Multiplatform stabilization we are trying to emphasize the essential and reduce ambiguous APIs. Therefore we discourage declaring multiple targets of the same kind in a single Gradle project.

The first sentence can also be dropped as it is just a "soft" introduction to what exactly we changed. But it can blur the reader's eye.

Comment (dsavvinov) [reply docs/topics/multiplatform/multiplatform-compatibility-guide.md]: > The first sentence can also be dropped

Yes, I'd agree

Comment (antohaby) [reply docs/topics/multiplatform/multiplatform-compatibility-guide.md]: Updated

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md (hunk 3 above)
Comment (antohaby) [reply docs/topics/multiplatform/multiplatform-compatibility-guide.md:688]: Updated

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md (hunk 4 above)
Comment (antohaby) [reply docs/topics/multiplatform/multiplatform-compatibility-guide.md]: Separated and I made more detailed samples. Can you check them?

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md (hunk 5 above)
Comment (antohaby) [reply docs/topics/multiplatform/multiplatform-compatibility-guide.md]: Reworked with tables. Tried to add detailed samples. I'm open for more suggetsions

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 6]
```diff
@@ -490,3 +490,67 @@ projects.
...
+```
+
+Different projects can be configured like this:
+```kotlin
```
[the remark this answers is not in the document]
Comment (antohaby) [reply docs/topics/multiplatform/multiplatform-compatibility-guide.md:645]: I've made a sample with UI library where you can have 'compose' and 'swing'. With some shared code between them.

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 7]
```diff
@@ -490,3 +490,200 @@ projects.
...
+    If you need to compile some sources with different parameters such as dependencies or compiler options for example
+    for functional tests. Then instead of declaring multiple targets you can create multiple compilations in one target.
+ 
+   <tabs group="build-script">
+   <tab title="Before" group-key="before">
+
+    ```kotlin
+   // Project ':myLibrary:integrationTests'
+   kotlin {
+      jvm("jvmKtor") // !!!
```
Comment (dsavvinov) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:525]: I'd recommend to avoid using JVM-shared source sets without the specific the need. I.e., let's add `linuxX64` or something here.

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 8]
```diff
@@ -490,3 +490,200 @@ projects.
...
+   
+    ```kotlin
+   // Project ':myLibrary:integrationTests'
+   kotlin {
+      jvm {
+         val ktorIntegrationTest by compilations.creating
+         val okioIntegrationTest by compilations.creating
+   
+         // Associate with test compilation to see tests from commonTest
+         ktorIntegrationTest.associateWith(compilations.getByName("test"))
```
Comment (dsavvinov) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:565]: Is `associateWith` considered non-experimental for KMP? I thought it's not, and only `associateWith` in JVM is.

Comment (dsavvinov) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:565]: Why is it associated with `test`? Does it even work in IDE? 

I think the direct counterpart for the previous snippet would be to:

- associate new compilations with `main`
- connect link their default source sets with `dependsOn` to the `commonTest`

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 9]
```diff
@@ -490,3 +490,200 @@ projects.
...
+2. Extracting new target into a separate Gradle project
+
+   <tabs group="build-script">
+   <tab title="Before" group-key="before">
+
+   ```kotlin
+   // Project ':myUILibrary'
+   kotlin {
+       jvm("jvmCompose") // !!!
+       jvm("jvmSwing")   // !!!
```
Comment (dsavvinov) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:616]: Ditto about having `commonMain` as JVM-shared source set

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 10]
```diff
@@ -490,3 +490,200 @@ projects.
...
+   }
+   ```
+
+   </tab>
+   <tab title="Now" group-key="now">
+
+```kotlin
+// Project ':myUILibrary
+kotlin { 
+   jvm()
```
Comment (dsavvinov) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:648]: I'd also avoid promoting single-target projects

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 11]
```diff
@@ -490,3 +490,200 @@ projects.
...
+               }
+           }
+       }
+   }
+   ```
+
+   </tab>
+   <tab title="Now" group-key="now">
+
+```kotlin
```
Comment (dsavvinov) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:645]: I'm a bit conernced that this example is neither concise, nor complete. I think we have to make a firm decision: either we're providing complete examples, or we don't provide them at all and just assume that the users that had multiple same targets are advanced enough to handle this refactoring on their own.

I'm leaning towards the latter. Mostly because it's so hard to give a proper minimal migration example -- as it involves multiple files in the project, the only solution I see is an actual GitHub project with "migration commit", but IMO it's too much for this case (but we can expand it, if we see demand/questions)

**Review by dsavvinov — CHANGES_REQUESTED** [review]
Thanks for providing examples, and sorry for suggesting to drop this work, but I'm really concerned that they are not "complete", but still quite verbose. Let's discuss the idea of dropping them completley?

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md (hunk 11 above)
Comment (martinbonnin) [reply docs/topics/multiplatform/multiplatform-compatibility-guide.md:645]: Agree with @dsavvinov here. Also, it would be nice to include a paragraph about when to choose separate compilations over separate Gradle projects. Adding a separate Gradle project is usually more involved. Unless there's something fundamental it unlocks, I would even not mention it.

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 12]
```diff
@@ -490,3 +490,200 @@ projects.
...
+         val okioIntegrationTest by compilations.creating
+   
+         // Associate with test compilation to see tests from commonTest
+         ktorIntegrationTest.associateWith(compilations.getByName("test"))
+         okioIntegrationTest.associateWith(compilations.getByName("test"))
+   
+         // Create a Test Run to for each integration test
+         testRuns {
+            create("ktorIntegration") { setExecutionSourceFrom(ktorIntegrationTest) }
+            create("okioIntegration") { setExecutionSourceFrom(okioIntegrationTest) }
```
Comment (martinbonnin) [inline docs/topics/multiplatform/multiplatform-compatibility-guide.md:571]: Can I do this for native targets? The below doesn't work:

```kotlin
kotlin {
  macosArm64 {
    testRuns {
      // Unresolved reference
      setExecutionSourceFrom(compilation)
    }
  }
}
```

File: docs/topics/multiplatform/multiplatform-compatibility-guide.md [hunk 13]
```diff
@@ -490,3 +490,200 @@ projects.
...
+**What's changed?**
+
+We discourage declaring multiple targets of the same kind in a single Gradle project.
+
+Example:
+```kotlin
+kotlin {
+    jvm()
+    jvm("jvmKtor")
+    jvm("jvmOkio") // is not recommended and will produce deprecation warning
```
[the remark this answers is not in the document]
Comment (antohaby) [reply docs/topics/multiplatform/multiplatform-compatibility-guide.md:506]: right, we can remake it to just:

```
jvm()
jvm("jvmKtor") // is not recommended...
```

**Comment by dsavvinov** [comment]
Changes in this request will be merged as part of the https://github.com/JetBrains/kotlin-web-site/pull/3789 . I'm closing this one.
